Source/WebCore/Modules/permissions/PermissionStatus.cpp

```diff
@@ -31,48 +31,97 @@
+#include <wtf/HashMap.h>
 #include <wtf/IsoMallocInlines.h>
 #include <wtf/MainThread.h>
+#include <wtf/NeverDestroyed.h>
 
 namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(PermissionStatus);
 
-Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+static uint64_t nextMainThreadPermissionObserverIdentifier()
 {
-    auto status = adoptRef(*new PermissionStatus(context, state, descriptor));
+    static std::atomic<uint64_t> currentIdentifier;
+    return ++currentIdentifier;
+}
+
+static HashMap<uint64_t, std::unique_ptr<MainThreadPermissionObserver>>& allMainThreadPermissionObservers()
+{
+    static NeverDestroyed<HashMap<uint64_t, std::unique_ptr<MainThreadPermissionObserver>>> map;
+    return map;
+}
+
+static void addMainThreadPermissionObserver(uint64_t identifier, std::unique_ptr<MainThreadPermissionObserver>&& observer)
+{
+    allMainThreadPermissionObservers().add(identifier, WTFMove(observer));
+}
+
+static void removeMainThreadPermissionObserver(uint64_t identifier)
+{
+    allMainThreadPermissionObservers().remove(identifier);
+}
+
+Ref<PermissionStatus> PermissionStatus::create(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, Page* page)
+{
+    auto status = adoptRef(*new PermissionStatus(context, state, descriptor, source, page));
+    if (source != PermissionQuerySource::Window)
+        status->createAssociatedMainThreadPermissionObserver(context);
     status->suspendIfNeeded();
     return status;
 }
 
-PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor)
+PermissionStatus::PermissionStatus(ScriptExecutionContext& context, PermissionState state, const PermissionDescriptor& descriptor, PermissionQuerySource source, Page* page)
     : ActiveDOMObject(&context)
     , m_state(state)
     , m_descriptor(descriptor)
+    , m_source(source)
+    , m_page(page)
 {
     auto* origin = context.securityOrigin();
     auto originData = origin ? origin->data() : SecurityOriginData { };
     m_origin = ClientOrigin { context.topOrigin().data(), WTFMove(originData) };
     
-    // FIXME: Add support for workers. Currently, this only works for Window objects.
-    if (isMainThread())
+    m_mainThreadPermissionObserverIdentifier = 0;
+    if (source == PermissionQuerySource::Window) {
         PermissionController::shared().addObserver(*this);
+        return;
+    }
+
+    m_mainThreadPermissionObserverIdentifier = nextMainThreadPermissionObserverIdentifier();
+}
+
+void PermissionStatus::createAssociatedMainThreadPermissionObserver(ScriptExecutionContext& context)
+{
+    auto& workerGlobalScope = downcast<WorkerGlobalScope>(context);
+    callOnMainThread([this, protectedThis = Ref { *this }, contextIdentifier = workerGlobalScope.identifier(), origin = m_origin.isolatedCopy()]() {
```

Comment:
Capturing protectedThis here means that the PermissionStatus object could now get destroyed on the main thread, even if it is a worker thread object. This is not correct and can lead to crashes.

One way to avoid it would be to re-dispatch protectedThis back to the contextThread once we're done with it on the main thread.

Reply:
I am not sure it will actually work since the context thread might be stopped at the time we are trying to hop it back to it to do the unref.
Maybe the object that got ref and passed to main thread should not be an ActiveDOMObject. and PermissionStatus would keep a ref to it.